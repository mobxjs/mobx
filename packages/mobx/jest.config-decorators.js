const path = require("path")
const buildConfig = require("../../jest.base.config")

module.exports = buildConfig(
    __dirname,
    {
        testRegex: "__tests__/decorators_20223/.*\\.(t|j)sx?$",
        setupFilesAfterEnv: [`<rootDir>/jest.setup.ts`]
    },
    path.resolve(__dirname, "./__tests__/decorators_20223/tsconfig.json")
)
