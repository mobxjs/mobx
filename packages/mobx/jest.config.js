const buildConfig = require("../../jest.base.config")

module.exports = buildConfig(__dirname, {
    projects: ["<rootDir>/jest.config.js", "<rootDir>/jest.config-decorators.js"],
    testRegex: "__tests__/v[4|5]/base/.*\\.(t|j)sx?$",
    setupFilesAfterEnv: [`<rootDir>/jest.setup.ts`]
})
