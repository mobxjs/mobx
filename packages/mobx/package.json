{
    "name": "mobx",
    "version": "6.13.7",
    "description": "Simple, scalable state management.",
    "source": "src/mobx.ts",
    "main": "dist/index.js",
    "umd:main": "dist/mobx.umd.production.min.js",
    "unpkg": "dist/mobx.umd.production.min.js",
    "jsdelivr": "dist/mobx.umd.production.min.js",
    "jsnext:main": "dist/mobx.esm.production.min.js",
    "module": "dist/mobx.esm.js",
    "react-native": "dist/mobx.esm.js",
    "types": "dist/mobx.d.ts",
    "typings": "dist/mobx.d.ts",
    "files": [
        "src",
        "dist",
        "LICENSE",
        "CHANGELOG.md",
        "README.md"
    ],
    "sideEffects": false,
    "repository": {
        "type": "git",
        "url": "https://github.com/mobxjs/mobx.git"
    },
    "author": "Michel Weststrate",
    "license": "MIT",
    "funding": {
        "type": "opencollective",
        "url": "https://opencollective.com/mobx"
    },
    "bugs": {
        "url": "https://github.com/mobxjs/mobx/issues"
    },
    "homepage": "https://mobx.js.org/",
    "dependencies": {},
    "devDependencies": {
        "@babel/core": "^7.9.0",
        "@babel/plugin-proposal-class-properties": "^7.8.3",
        "@babel/plugin-proposal-decorators": "^7.8.3",
        "@babel/plugin-transform-runtime": "^7.9.0",
        "@babel/preset-env": "^7.9.5",
        "@babel/preset-typescript": "^7.9.0",
        "@babel/runtime": "^7.9.2",
        "conditional-type-checks": "^1.0.5",
        "flow-bin": "^0.123.0"
    },
    "keywords": [
        "mobx",
        "mobservable",
        "observable",
        "react-component",
        "react",
        "reactjs",
        "reactive",
        "model",
        "frp",
        "functional-reactive-programming",
        "state management",
        "data flow"
    ],
    "scripts": {
        "test": "jest",
        "lint": "eslint src/**/*",
        "build": "node ../../scripts/build.js mobx",
        "build:test": "yarn build --target test",
        "perf": "scripts/perf.sh",
        "perf-legacy": "node --expose-gc ./__tests__/perf/index.js legacy",
        "perf-proxy": "node --expose-gc ./__tests__/perf/index.js proxy",
        "test:performance": "yarn perf proxy && yarn perf legacy",
        "test:mixed-versions": "yarn test --testRegex mixed-versions",
        "test:types": "tsc --noEmit",
        "test:flow": "flow check",
        "test:coverage": "yarn test -i --coverage",
        "test:size": "yarn import-size --report . observable computed autorun action",
        "test:check": "yarn test:types",
        "prepublishOnly": "node ./scripts/prepublish.js && yarn build --target publish"
    }
}
