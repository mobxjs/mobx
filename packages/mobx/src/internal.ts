/*
The only reason for this file to exist is pure horror:
Without it rollup can make the bundling fail at any point in time; when it rolls up the files in the wrong order
it will cause undefined errors (for example because super classes or local variables not being hoisted).
With this file that will still happen,
but at least in this file we can magically reorder the imports with trial and error until the build succeeds again.
*/
export * from "./utils/global"
export * from "./errors"
export * from "./utils/utils"
export * from "./utils/weakset"
export * from "./api/decorators"
export * from "./core/atom"
export * from "./utils/comparer"
export * from "./types/modifiers"
export * from "./types/overrideannotation"
export * from "./types/actionannotation"
export * from "./types/flowannotation"
export * from "./types/computedannotation"
export * from "./types/observableannotation"
export * from "./types/autoannotation"
export * from "./types/generic-abort-signal"
export * from "./api/observable"
export * from "./api/computed"
export * from "./core/action"
export * from "./types/observablevalue"
export * from "./core/computedvalue"
export * from "./core/derivation"
export * from "./core/globalstate"
export * from "./core/observable"
export * from "./core/reaction"
export * from "./core/spy"
export * from "./api/action"
export * from "./api/autorun"
export * from "./api/become-observed"
export * from "./api/configure"
export * from "./api/extendobservable"
export * from "./api/extras"
export * from "./api/flow"
export * from "./api/intercept-read"
export * from "./api/intercept"
export * from "./api/iscomputed"
export * from "./api/isobservable"
export * from "./api/object-api"
export * from "./api/observe"
export * from "./api/tojs"
export * from "./api/trace"
export * from "./api/transaction"
export * from "./api/when"
export * from "./types/dynamicobject"
export * from "./types/intercept-utils"
export * from "./types/listen-utils"
export * from "./api/makeObservable"
export * from "./types/observablearray"
export * from "./types/observablemap"
export * from "./types/observableset"
export * from "./types/observableobject"
export * from "./types/legacyobservablearray"
export * from "./types/type-utils"
export * from "./utils/eq"
export * from "./utils/iterable"
export * from "./api/annotation"
