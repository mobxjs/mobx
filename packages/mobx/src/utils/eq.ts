import {
    isES6Map,
    isObservableArray,
    isObservableMap,
    isES6Set,
    isObservableSet,
    hasProp,
    isFunction,
    objectPrototype
} from "../internal"

declare const Symbol
const toString = objectPrototype.toString

export function deepEqual(a: any, b: any, depth: number = -1): boolean {
    return eq(a, b, depth)
}

// Copied from https://github.com/jashkenas/underscore/blob/5c237a7c682fb68fd5378203f0bf22dce1624854/underscore.js#L1186-L1289
// Modified: "Deep compare objects" part to iterate over keys in forward order instead of reverse order.
//
// Internal recursive comparison function for `isEqual`.
function eq(a: any, b: any, depth: number, aStack?: any[], bStack?: any[]) {
    // Identical objects are equal. `0 === -0`, but they aren't identical.
    // See the [Harmony `egal` proposal](http://wiki.ecmascript.org/doku.php?id=harmony:egal).
    if (a === b) {
        return a !== 0 || 1 / a === 1 / b
    }
    // `null` or `undefined` only equal to itself (strict comparison).
    if (a == null || b == null) {
        return false
    }
    // `NaN`s are equivalent, but non-reflexive.
    if (a !== a) {
        return b !== b
    }
    // Exhaust primitive checks
    const type = typeof a
    if (type !== "function" && type !== "object" && typeof b != "object") {
        return false
    }

    // Compare `[[Class]]` names.
    const className = toString.call(a)
    if (className !== toString.call(b)) {
        return false
    }
    switch (className) {
        // Strings, numbers, regular expressions, dates, and booleans are compared by value.
        case "[object RegExp]":
        // RegExps are coerced to strings for comparison (Note: '' + /a/i === '/a/i')
        case "[object String]":
            // Primitives and their corresponding object wrappers are equivalent; thus, `"5"` is
            // equivalent to `new String("5")`.
            return "" + a === "" + b
        case "[object Number]":
            // `NaN`s are equivalent, but non-reflexive.
            // Object(NaN) is equivalent to NaN.
            if (+a !== +a) {
                return +b !== +b
            }
            // An `egal` comparison is performed for other numeric values.
            return +a === 0 ? 1 / +a === 1 / b : +a === +b
        case "[object Date]":
        case "[object Boolean]":
            // Coerce dates and booleans to numeric primitive values. Dates are compared by their
            // millisecond representations. Note that invalid dates with millisecond representations
            // of `NaN` are not equivalent.
            return +a === +b
        case "[object Symbol]":
            return (
                typeof Symbol !== "undefined" && Symbol.valueOf.call(a) === Symbol.valueOf.call(b)
            )
        case "[object Map]":
        case "[object Set]":
            // Maps and Sets are unwrapped to arrays of entry-pairs, adding an incidental level.
            // Hide this extra level by increasing the depth.
            if (depth >= 0) {
                depth++
            }
            break
    }
    // Unwrap any wrapped objects.
    a = unwrap(a)
    b = unwrap(b)

    const areArrays = className === "[object Array]"
    if (!areArrays) {
        if (typeof a != "object" || typeof b != "object") {
            return false
        }

        // Objects with different constructors are not equivalent, but `Object`s or `Array`s
        // from different frames are.
        const aCtor = a.constructor,
            bCtor = b.constructor
        if (
            aCtor !== bCtor &&
            !(
                isFunction(aCtor) &&
                aCtor instanceof aCtor &&
                isFunction(bCtor) &&
                bCtor instanceof bCtor
            ) &&
            "constructor" in a &&
            "constructor" in b
        ) {
            return false
        }
    }

    if (depth === 0) {
        return false
    } else if (depth < 0) {
        depth = -1
    }

    // Assume equality for cyclic structures. The algorithm for detecting cyclic
    // structures is adapted from ES 5.1 section 15.12.3, abstract operation `JO`.

    // Initializing stack of traversed objects.
    // It's done here since we only need them for objects and arrays comparison.
    aStack = aStack || []
    bStack = bStack || []
    let length = aStack.length
    while (length--) {
        // Linear search. Performance is inversely proportional to the number of
        // unique nested structures.
        if (aStack[length] === a) {
            return bStack[length] === b
        }
    }

    // Add the first object to the stack of traversed objects.
    aStack.push(a)
    bStack.push(b)

    // Recursively compare objects and arrays.
    if (areArrays) {
        // Compare array lengths to determine if a deep comparison is necessary.
        length = a.length
        if (length !== b.length) {
            return false
        }
        // Deep compare the contents, ignoring non-numeric properties.
        while (length--) {
            if (!eq(a[length], b[length], depth - 1, aStack, bStack)) {
                return false
            }
        }
    } else {
        // Deep compare objects.
        const keys = Object.keys(a)
        const length = keys.length
        // Ensure that both objects contain the same number of properties before comparing deep equality.
        if (Object.keys(b).length !== length) {
            return false
        }
        for (let i = 0; i < length; i++) {
            // Deep compare each member
            const key = keys[i]
            if (!(hasProp(b, key) && eq(a[key], b[key], depth - 1, aStack, bStack))) {
                return false
            }
        }
    }
    // Remove the first object from the stack of traversed objects.
    aStack.pop()
    bStack.pop()
    return true
}

function unwrap(a: any) {
    if (isObservableArray(a)) {
        return a.slice()
    }
    if (isES6Map(a) || isObservableMap(a)) {
        return Array.from(a.entries())
    }
    if (isES6Set(a) || isObservableSet(a)) {
        return Array.from(a.entries())
    }
    return a
}
