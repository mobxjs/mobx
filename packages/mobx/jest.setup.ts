import { configure, _resetGlobalState } from "./src/mobx"

global.setImmediate = global.setImmediate || ((fn, ...args) => global.setTimeout(fn, 0, ...args))

beforeEach(() => {
    // @ts-ignore
    global.__DEV__ = true
    _resetGlobalState()
    configure({
        enforceActions: "never"
    })
})
