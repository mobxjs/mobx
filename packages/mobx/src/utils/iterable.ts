import { getGlobal } from "../internal"

// safely get iterator prototype if available
const maybeIteratorPrototype = getGlobal().Iterator?.prototype || {}

export function makeIterable<T, TReturn = unknown>(
    iterator: Iterator<T>
): IteratorObject<T, TReturn> {
    iterator[Symbol.iterator] = getSelf
    return Object.assign(Object.create(maybeIteratorPrototype), iterator)
}

function getSelf() {
    return this
}
