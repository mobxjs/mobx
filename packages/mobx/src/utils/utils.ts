import { globalState, die } from "../internal"

// We shorten anything used > 5 times
export const assign = Object.assign
export const getDescriptor = Object.getOwnPropertyDescriptor
export const defineProperty = Object.defineProperty
export const objectPrototype = Object.prototype

export const EMPTY_ARRAY = []
Object.freeze(EMPTY_ARRAY)

export const EMPTY_OBJECT = {}
Object.freeze(EMPTY_OBJECT)

export interface Lambda {
    (): void
    name?: string
}

const hasProxy = typeof Proxy !== "undefined"
const plainObjectString = Object.toString()

export function assertProxies() {
    if (!hasProxy) {
        die(
            __DEV__
                ? "`Proxy` objects are not available in the current environment. Please configure MobX to enable a fallback implementation.`"
                : "Proxy not available"
        )
    }
}

export function warnAboutProxyRequirement(msg: string) {
    if (__DEV__ && globalState.verifyProxies) {
        die(
            "MobX is currently configured to be able to run in ES5 mode, but in ES5 MobX won't be able to " +
                msg
        )
    }
}

export function getNextId() {
    return ++globalState.mobxGuid
}

/**
 * Makes sure that the provided function is invoked at most once.
 */
export function once(func: Lambda): Lambda {
    let invoked = false
    return function () {
        if (invoked) {
            return
        }
        invoked = true
        return (func as any).apply(this, arguments)
    }
}

export const noop = () => {}

export function isFunction(fn: any): fn is Function {
    return typeof fn === "function"
}

export function isString(value: any): value is string {
    return typeof value === "string"
}

export function isStringish(value: any): value is string | number | symbol {
    const t = typeof value
    switch (t) {
        case "string":
        case "symbol":
        case "number":
            return true
    }
    return false
}

export function isObject(value: any): value is Object {
    return value !== null && typeof value === "object"
}

export function isPlainObject(value: any) {
    if (!isObject(value)) {
        return false
    }
    const proto = Object.getPrototypeOf(value)
    if (proto == null) {
        return true
    }
    const protoConstructor = Object.hasOwnProperty.call(proto, "constructor") && proto.constructor
    return (
        typeof protoConstructor === "function" && protoConstructor.toString() === plainObjectString
    )
}

// https://stackoverflow.com/a/37865170
export function isGenerator(obj: any): boolean {
    const constructor = obj?.constructor
    if (!constructor) {
        return false
    }
    if (
        "GeneratorFunction" === constructor.name ||
        "GeneratorFunction" === constructor.displayName
    ) {
        return true
    }
    return false
}

export function addHiddenProp(object: any, propName: PropertyKey, value: any) {
    defineProperty(object, propName, {
        enumerable: false,
        writable: true,
        configurable: true,
        value
    })
}

export function addHiddenFinalProp(object: any, propName: PropertyKey, value: any) {
    defineProperty(object, propName, {
        enumerable: false,
        writable: false,
        configurable: true,
        value
    })
}

export function createInstanceofPredicate<T>(
    name: string,
    theClass: new (...args: any[]) => T
): (x: any) => x is T {
    const propName = "isMobX" + name
    theClass.prototype[propName] = true
    return function (x) {
        return isObject(x) && x[propName] === true
    } as any
}

/**
 * Yields true for both native and observable Map, even across different windows.
 */
export function isES6Map(thing: unknown): thing is Map<any, any> {
    return thing != null && Object.prototype.toString.call(thing) === "[object Map]"
}

/**
 * Makes sure a Map is an instance of non-inherited native or observable Map.
 */
export function isPlainES6Map(thing: Map<unknown, unknown>): boolean {
    const mapProto = Object.getPrototypeOf(thing)
    const objectProto = Object.getPrototypeOf(mapProto)
    const nullProto = Object.getPrototypeOf(objectProto)
    return nullProto === null
}

/**
 * Yields true for both native and observable Set, even across different windows.
 */
export function isES6Set(thing: unknown): thing is Set<any> {
    return thing != null && Object.prototype.toString.call(thing) === "[object Set]"
}

const hasGetOwnPropertySymbols = typeof Object.getOwnPropertySymbols !== "undefined"

/**
 * Returns the following: own enumerable keys and symbols.
 */
export function getPlainObjectKeys(object: any) {
    const keys = Object.keys(object)
    // Not supported in IE, so there are not going to be symbol props anyway...
    if (!hasGetOwnPropertySymbols) {
        return keys
    }
    const symbols = Object.getOwnPropertySymbols(object)
    if (!symbols.length) {
        return keys
    }
    return [...keys, ...symbols.filter(s => objectPrototype.propertyIsEnumerable.call(object, s))]
}

// From Immer utils
// Returns all own keys, including non-enumerable and symbolic
export const ownKeys: (target: any) => Array<string | symbol> =
    typeof Reflect !== "undefined" && Reflect.ownKeys
        ? Reflect.ownKeys
        : hasGetOwnPropertySymbols
        ? obj => Object.getOwnPropertyNames(obj).concat(Object.getOwnPropertySymbols(obj) as any)
        : /* istanbul ignore next */ Object.getOwnPropertyNames

export function stringifyKey(key: any): string {
    if (typeof key === "string") {
        return key
    }
    if (typeof key === "symbol") {
        return key.toString()
    }
    return new String(key).toString()
}

export function toPrimitive(value: any) {
    return value === null ? null : typeof value === "object" ? "" + value : value
}

export function hasProp(target: Object, prop: PropertyKey): boolean {
    return objectPrototype.hasOwnProperty.call(target, prop)
}

// From Immer utils
export const getOwnPropertyDescriptors =
    Object.getOwnPropertyDescriptors ||
    function getOwnPropertyDescriptors(target: any) {
        // Polyfill needed for Hermes and IE, see https://github.com/facebook/hermes/issues/274
        const res: any = {}
        // Note: without polyfill for ownKeys, symbols won't be picked up
        ownKeys(target).forEach(key => {
            res[key] = getDescriptor(target, key)
        })
        return res
    }

export function getFlag(flags: number, mask: number) {
    return !!(flags & mask)
}

export function setFlag(flags: number, mask: number, newValue: boolean): number {
    if (newValue) {
        flags |= mask
    } else {
        flags &= ~mask
    }
    return flags
}
