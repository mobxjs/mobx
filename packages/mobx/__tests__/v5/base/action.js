import * as mobx from "../../../src/mobx"
import * as utils from "../utils/test-utils"

test("action should wrap in transaction", () => {
    const values = []

    const observable = mobx.observable.box(0)
    mobx.autorun(() => values.push(observable.get()))

    const increment = mobx.action("increment", amount => {
        observable.set(observable.get() + amount * 2)
        observable.set(observable.get() - amount) // oops
    })

    expect(mobx.isAction(increment)).toBe(true)
    expect(mobx.isAction(function () {})).toBe(false)

    increment(7)

    expect(values).toEqual([0, 7])
})

test("action modifications should be picked up 1", () => {
    const a = mobx.observable.box(1)
    let i = 3
    let b = 0

    mobx.autorun(() => {
        b = a.get() * 2
    })

    expect(b).toBe(2)

    const action = mobx.action(() => {
        a.set(++i)
    })

    action()
    expect(b).toBe(8)

    action()
    expect(b).toBe(10)
})

test("action modifications should be picked up 1", () => {
    const a = mobx.observable.box(1)
    let b = 0

    mobx.autorun(() => {
        b = a.get() * 2
    })

    expect(b).toBe(2)

    const action = mobx.action(() => {
        a.set(a.get() + 1) // ha, no loop!
    })

    action()
    expect(b).toBe(4)

    action()
    expect(b).toBe(6)
})

test("action modifications should be picked up 3", () => {
    const a = mobx.observable.box(1)
    let b = 0

    const doubler = mobx.computed(() => a.get() * 2)

    mobx.observe(
        doubler,
        () => {
            b = doubler.get()
        },
        true
    )

    expect(b).toBe(2)

    const action = mobx.action(() => {
        a.set(a.get() + 1) // ha, no loop!
    })

    action()
    expect(b).toBe(4)

    action()
    expect(b).toBe(6)
})

test("test action should be untracked", () => {
    const a = mobx.observable.box(3)
    const b = mobx.observable.box(4)
    let latest = 0
    let runs = 0

    const action = mobx.action(baseValue => {
        b.set(baseValue * 2)
        latest = b.get() // without action this would trigger loop
    })

    const d = mobx.autorun(() => {
        runs++
        const current = a.get()
        action(current)
    })

    expect(b.get()).toBe(6)
    expect(latest).toBe(6)

    a.set(7)
    expect(b.get()).toBe(14)
    expect(latest).toBe(14)

    a.set(8)
    expect(b.get()).toBe(16)
    expect(latest).toBe(16)

    b.set(7) // should have no effect
    expect(a.get()).toBe(8)
    expect(b.get()).toBe(7)
    expect(latest).toBe(16) // effect not triggered

    a.set(3)
    expect(b.get()).toBe(6)
    expect(latest).toBe(6)

    expect(runs).toBe(4)

    d()
})

test("should be possible to create autorun in action", () => {
    const a = mobx.observable.box(1)
    const values = []

    const adder = mobx.action(inc => {
        return mobx.autorun(() => {
            values.push(a.get() + inc)
        })
    })

    const d1 = adder(2)
    a.set(3)
    const d2 = adder(17)
    a.set(24)
    d1()
    a.set(11)
    d2()
    a.set(100)

    expect(values).toEqual([3, 5, 20, 26, 41, 28]) // n.b. order could swap as autorun creation order doesn't guarantee stuff
})

test("should be possible to change unobserved state in an action called from computed", () => {
    const a = mobx.observable.box(2)

    const testAction = mobx.action(() => {
        a.set(3)
    })

    const c = mobx.computed(() => {
        testAction()
    })

    expect.assertions(1)
    mobx.autorun(() => {
        expect(() => {
            c.get()
        }).not.toThrow(/bla/)
    })

    mobx._resetGlobalState()
})

test("should be possible to change observed state in an action called from computed", () => {
    const a = mobx.observable.box(2)
    const d = mobx.autorun(() => {
        a.get()
    })

    const testAction = mobx.action(() => {
        a.set(5) // this is fine
        expect(a.get()).toBe(5)
    })

    const c = mobx.computed(() => {
        expect(
            utils.grabConsole(() => {
                a.set(4)
            })
        ).toMatchInlineSnapshot(
            `"<STDOUT> [MobX] Side effects like changing state are not allowed at this point. Are you trying to modify state from, for example, a computed value or the render function of a React component? You can wrap side effects in 'runInAction' (or decorate functions with 'action') if needed. Tried to modify: ObservableValue@19"`
        )
        expect(a.get()).toBe(4)
        testAction()
        return a.get()
    })

    expect(c.get()).toBe(5)

    mobx._resetGlobalState()
    d()
})

test("should be possible to change observed state in an action called from computed", () => {
    const a = mobx.observable.box(2)
    const d = mobx.autorun(() => {
        a.get()
    })

    const testAction = mobx.action(() => {
        a.set(3)
    })

    const c = mobx.computed(() => {
        testAction()
        return a.get()
    })

    expect(
        utils.grabConsole(() => {
            c.get()
        })
    ).toBe("")

    mobx._resetGlobalState()
    d()
})

test("action in autorun should be untracked", () => {
    const a = mobx.observable.box(2)
    const b = mobx.observable.box(3)

    const data = []
    const multiplier = mobx.action(val => val * b.get())

    const d = mobx.autorun(() => {
        data.push(multiplier(a.get()))
    })

    a.set(3)
    b.set(4)
    a.set(5)

    d()

    a.set(6)

    expect(data).toEqual([6, 9, 20])
})

test("action should not be converted to computed when using (extend)observable", () => {
    const a = mobx.observable({
        a: 1,
        b: mobx.action(function () {
            this.a++
        })
    })

    expect(mobx.isAction(a.b)).toBe(true)
    a.b()
    expect(a.a).toBe(2)

    mobx.extendObservable(a, {
        c: mobx.action(function () {
            this.a *= 3
        })
    })

    expect(mobx.isAction(a.c)).toBe(true)
    a.c()
    expect(a.a).toBe(6)
})

test("#286 exceptions in actions should not affect global state", () => {
    let autorunTimes = 0
    function Todos() {
        mobx.extendObservable(this, {
            count: 0,
            add: mobx.action(function () {
                this.count++
                if (this.count === 2) {
                    throw new Error("An Action Error!")
                }
            })
        })
    }
    const todo = new Todos()
    mobx.autorun(() => {
        autorunTimes++
        return todo.count
    })
    try {
        todo.add()
        expect(autorunTimes).toBe(2)
        todo.add()
    } catch (e) {
        expect(autorunTimes).toBe(3)
        todo.add()
        expect(autorunTimes).toBe(4)
    }
})

test("runInAction", () => {
    mobx.configure({ enforceActions: "observed" })
    const values = []
    const events = []
    const spyDisposer = mobx.spy(ev => {
        if (ev.type === "action")
            events.push({
                name: ev.name,
                arguments: ev.arguments
            })
    })

    const observable = mobx.observable.box(0)
    const d = mobx.autorun(() => values.push(observable.get()))

    let res = mobx.runInAction(() => {
        observable.set(observable.get() + 6 * 2)
        observable.set(observable.get() - 3) // oops
        return 2
    })

    expect(res).toBe(2)
    expect(values).toEqual([0, 9])

    res = mobx.runInAction(() => {
        observable.set(observable.get() + 5 * 2)
        observable.set(observable.get() - 4) // oops
        return 3
    })

    expect(res).toBe(3)
    expect(values).toEqual([0, 9, 15])
    expect(events).toEqual([
        { arguments: [], name: "<unnamed action>" },
        { arguments: [], name: "<unnamed action>" }
    ])

    mobx.configure({ enforceActions: "never" })
    spyDisposer()

    d()
})

test("action in autorun does not keep / make computed values alive", () => {
    let calls = 0
    const myComputed = mobx.computed(() => calls++)
    const callComputedTwice = () => {
        myComputed.get()
        myComputed.get()
    }

    const runWithMemoizing = fun => {
        mobx.autorun(fun)()
    }

    callComputedTwice()
    expect(calls).toBe(2)

    runWithMemoizing(callComputedTwice)
    expect(calls).toBe(3)

    callComputedTwice()
    expect(calls).toBe(5)

    runWithMemoizing(function () {
        mobx.runInAction(callComputedTwice)
    })
    expect(calls).toBe(6)

    callComputedTwice()
    expect(calls).toBe(8)
})

test("computed values and actions", () => {
    let calls = 0

    const number = mobx.observable.box(1)
    const squared = mobx.computed(() => {
        calls++
        return number.get() * number.get()
    })
    const changeNumber10Times = mobx.action(() => {
        squared.get()
        squared.get()
        for (let i = 0; i < 10; i++) number.set(number.get() + 1)
    })

    changeNumber10Times()
    expect(calls).toBe(1)

    mobx.autorun(() => {
        changeNumber10Times()
        expect(calls).toBe(2)
    })()
    expect(calls).toBe(2)

    changeNumber10Times()
    expect(calls).toBe(3)
})

test("extendObservable respects action decorators", () => {
    const x = mobx.observable(
        {
            a1() {
                return this
            },
            a2() {
                return this
            },
            a3() {
                return this
            }
        },
        {
            a1: mobx.action,
            a2: mobx.action.bound,
            a3: false
        }
    )
    expect(mobx.isAction(x.a1)).toBe(true)
    expect(mobx.isAction(x.a2)).toBe(true)
    expect(mobx.isAction(x.a3)).toBe(false)

    // const global = (function() {
    //     return this
    // })()

    const { a1, a2, a3 } = x
    expect(a1.call(x)).toBe(x)
    // expect(a1()).toBe(global)
    expect(a2.call(x)).toBeTruthy() // it is not this! proxies :) see test in proxies.js
    expect(a2()).toBeTruthy()
    expect(a3.call(x)).toBe(x)
    // expect(a3()).toBe(global)
})

test("expect warning for invalid decorator", () => {
    expect(() => {
        mobx.observable({ x: 1 }, { x: undefined })
    }).toThrow(/Invalid annotation/)
})

test("bound actions bind", () => {
    let called = 0
    const x = mobx.observable(
        {
            y: 0,
            z: function (v) {
                this.y += v
                this.y += v
            },
            get yValue() {
                called++
                return this.y
            }
        },
        {
            z: mobx.action.bound
        }
    )

    const d = mobx.autorun(() => {
        x.yValue
    })
    const events = []
    const d2 = mobx.spy(e => events.push(e))

    const runner = x.z
    runner(3)
    expect(x.yValue).toBe(6)
    expect(called).toBe(2)

    expect(events.filter(e => e.type === "action").map(e => e.name)).toEqual(["z"])
    expect(Object.keys(x)).toEqual(["y"])

    d()
    d2()
})

test("Fix #1367", () => {
    const x = mobx.extendObservable(
        {},
        {
            method() {}
        },
        {
            method: mobx.action
        }
    )
    expect(mobx.isAction(x.method)).toBe(true)
})

test("error logging, #1836 - 1", () => {
    const messages = utils.supressConsole(() => {
        try {
            const a = mobx.observable.box(3)
            mobx.autorun(() => {
                if (a.get() === 4) throw new Error("Reaction error")
            })

            mobx.action(() => {
                a.set(4)
                throw new Error("Action error")
            })()
        } catch (e) {
            expect(e.toString()).toEqual("Error: Action error")
            console.error(e)
        }
    })

    expect(messages).toMatchSnapshot()
})

test("error logging, #1836 - 2", () => {
    const messages = utils.supressConsole(() => {
        try {
            const a = mobx.observable.box(3)
            mobx.autorun(() => {
                if (a.get() === 4) throw new Error("Reaction error")
            })

            mobx.action(() => {
                a.set(4)
            })()
        } catch (e) {
            expect(e.toString()).toEqual("Error: Action error")
            console.error(e)
        }
    })

    expect(messages).toMatchSnapshot()
})

test("out of order startAction / endAction", () => {
    const a1 = mobx._startAction("a1")
    const a2 = mobx._startAction("a2")

    expect(() => mobx._endAction(a1)).toThrow("invalid action stack")

    mobx._endAction(a2)

    // double finishing
    expect(() => mobx._endAction(a2)).toThrow("invalid action stack")

    mobx._endAction(a1)
})

test("given actionName, the action function name should be defined as the actionName", () => {
    const a1 = mobx.action("testAction", () => {})
    expect(a1.name).toBe("testAction")
})

test("given anonymous action, the action name should be <unnamed action>", () => {
    const a1 = mobx.action(() => {})
    expect(a1.name).toBe("<unnamed action>")
})

test("given function declaration, the action name should be as the function name", () => {
    const a1 = mobx.action(function testAction() {})
    expect(a1.name).toBe("testAction")
})

test("auto action can be used in a derivation and is tracked", () => {
    const a = mobx.observable(1)
    const events = []

    const double = mobx._autoAction(() => {
        return a.get() * 2
    })

    const d = mobx.autorun(() => {
        events.push(double())
    })

    a.set(2)
    expect(events).toEqual([2, 4])
    d()
})

test("auto action can be used to update and is batched", () => {
    const a = mobx.observable(1)
    const events = []

    const d = mobx.autorun(() => {
        events.push(a.get() * 2)
    })

    mobx._autoAction(() => {
        a.set(2)
        a.set(3)
    })()

    expect(events).toEqual([2, 6]) // No 4!
    d()
})

test("auto action should not update state from inside a derivation", async () => {
    const a = mobx.observable(1)

    const d = mobx.autorun(() => a.get()) // observe

    const double = mobx._autoAction(() => {
        a.set(a.get() * 2)
    })

    await mobx.when(() => {
        expect(
            utils.grabConsole(() => {
                double()
            })
        ).toMatchInlineSnapshot(
            `"<STDOUT> [MobX] Side effects like changing state are not allowed at this point. Are you trying to modify state from, for example, a computed value or the render function of a React component? You can wrap side effects in 'runInAction' (or decorate functions with 'action') if needed. Tried to modify: ObservableValue@51"`
        )
        return a.get() === 2
    })
    d()
})

test("auto action should not update state from inside a derivation", async () => {
    const a = mobx.observable(1)

    const d = mobx.autorun(() => a.get()) // observe

    const double = mobx._autoAction(() => {
        a.set(a.get() * 2)
    })

    await mobx.when(() => {
        expect(
            utils.grabConsole(() => {
                mobx.runInAction(() => {
                    // extra nesting, just in case
                    mobx._autoAction(() => {
                        double()
                    })()
                })
            })
        ).toBe("")
        return a.get() === 2
    })
    d()
})

test("action forwards toString of underlying function", async () => {
    const fn = () => {
        /* not actually doing anything */
    }
    fn.a = 42
    fn.toString = function () {
        return `toString referencing this, a=${this.a}`
    }

    const act = mobx.action(fn)

    expect(fn.toString()).toBe("toString referencing this, a=42")
    expect(act.toString()).toBe("toString referencing this, a=42")
})
