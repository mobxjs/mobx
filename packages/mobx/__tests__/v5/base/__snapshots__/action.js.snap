// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error logging, #1836 - 1 1`] = `
[
  "<STDOUT> [mobx] (error in reaction 'Autorun@44' suppressed, fix error of causing action below)",
  "<STDERR> Error: Action error",
]
`;

exports[`error logging, #1836 - 2 1`] = `
[
  "<STDERR> [mobx] Encountered an uncaught exception that was thrown by a reaction or observer component, in: 'Reaction[Autorun@46]'",
]
`;
